substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/SystemPropertiesSupport.java

```diff
@@ -48,25 +57,126 @@
+            this.sameAsHosted = sameAsHosted;
+            this.initValue = initValue;
+        }
+
+        public String name;
+        public int version;
+        public ChangedType changed;
+        public boolean sameAsHosted;
+        public String initValue;
+    }
+
     /** System properties that are taken from the VM hosting the image generator. */
-    private static final String[] HOSTED_PROPERTIES = {
-                    "java.version",
-                    ImageInfo.PROPERTY_IMAGE_KIND_KEY,
-                    /*
-                     * We do not support cross-compilation for now. Separator might also be cached
-                     * in other classes, so changing them would be tricky.
-                     */
-                    "line.separator", "path.separator", "file.separator",
-                    /* For our convenience for now. */
-                    "file.encoding", "sun.jnu.encoding",
-                    "java.class.version",
-                    "java.specification.name",
-                    "java.specification.vendor",
-                    "java.specification.version",
-                    "java.vm.specification.name",
-                    "java.vm.specification.vendor",
-                    "java.vm.specification.version"
-    };
+    private static final Set<String> HOSTED_PROPERTIES = new HashSet<>(Arrays.asList(
+            "java.version",
+            ImageInfo.PROPERTY_IMAGE_KIND_KEY,
+            /*
+             * We do not support cross-compilation for now. Separator might also be cached
+             * in other classes, so changing them would be tricky.
+             */
+            "line.separator", "path.separator", "file.separator",
+            /* For our convenience for now. */
+            "file.encoding", "sun.jnu.encoding",
+            "java.class.version",
+            "java.specification.name",
+            "java.specification.vendor",
+            "java.specification.version",
+            "java.vm.specification.name",
+            "java.vm.specification.vendor",
+            "java.vm.specification.version",
+            "java.runtime.version",
+            "sun.cpu.endian",
+            "sun.cpu.isalist",
+            "sun.io.unicode.encoding"));
+
+    /** Undefined properties that are not changed since JDK8.
+     * This collection is used as baseline*/
+    private static final String[] UNDEFINED_UNCHANGED_PROPERTIES_8 = {
+                    "awt.toolkit",
+                    "java.awt.graphicsenv",
+                    "java.awt.printerjob",
+                    "java.home",
+                    "sun.java.command",
+                    "sun.java.launcher",
+                    "sun.os.patch.level",
+                    "user.country",
+                    "user.language",
+                    "user.timezone"};
+
+    /** Properties are removed or added since JDK8*/
+    private static final List<PropertyInfo> CHANGED_PROPERTIES_SINCE_8 = new ArrayList<>(
```

Comment:
I don't like this "delta" list, since it is difficult to maintain in the long run. Instead, we should change the whole mechansim we have right now (`HOSTED_PROPERTIES` and `initializeProperty()`)  to a `PropertyInfo` like you introduced - without the need for "added" and "deleted" but instead with a predicate `onlyWith` where you can specify the existing classes like `JDK8OrEarlier` or `JDK11OrLater`.
Note also that we do not need to support only JDK 8, 11, and 15 - and not any versions in between.

Reply:
> Note also that we do not need to support only JDK 8, 11, and 15 - and not any versions in between.

Do you mean only support 8, 11 and 15, or all the versions in between? I think it should be the later. For example, `java.vm.compressedOopsMode` is introduced in JDK 9, its `onlyWith` should be `JDK9OrLater`.

Reply:
I mean "only support 8, 11 and 15". Many parts of native image don't support any of the versions in between already. Sticking to the long term support versions plus "latest" keeps the code base manageable.

Reply:
Refactored.